CRM/Contribute/BAO/Contribution.php

```diff
@@ -163,9 +163,12 @@ public static function add(&$params, $ids = array()) {
 
     $contribution->id = $contributionID;
 
-    if (!CRM_Utils_Rule::currencyCode($contribution->currency)) {
-      $config = CRM_Core_Config::singleton();
-      $contribution->currency = $config->defaultCurrency;
+    if ($contribution->id) {
+      // (only) on 'create', make sure that a valid currency is set (CRM-16845)
+      if (!CRM_Utils_Rule::currencyCode($contribution->currency)) {
+        $config = CRM_Core_Config::singleton();
+        $contribution->currency = $config->defaultCurrency;
```

Comment:
trivial aside you can now save a line by doing

CRM_Core_Config::singleton()->defaultCurrency;